l5kit/l5kit/simulation/unroll.py

```diff
@@ -0,0 +1,226 @@
+        """
+        Create a simulation loop object capable of unrolling ego and agents
+        :param sim_cfg: configuration for unroll
+        :param dataset: EgoDataset used while unrolling
+        :param model_ego: the model to be used for ego
+        :param model_agents: the model to be used for agents
+        """
+        self.sim_cfg = sim_cfg
+        if not sim_cfg.use_ego_gt and model_ego is None:
+            raise ValueError("ego model should not be None when simulating ego")
+        if not sim_cfg.use_agents_gt and model_agents is None:
+            raise ValueError("agents model should not be None when simulating agent")
+
+        self.model_ego = model_ego
+        self.model_agents = model_agents
+        self.dataset = dataset
+
+    def unroll(self, scene_indices: List[int]) -> List[SimulationOutputs]:
+        """
+        Simulate the dataset for the given scene indices
+        :param scene_indices: the scene indices we want to simulate
+        :return: the simulated dataset
+        TODO: this should probably return something else
+        """
+        sim_dataset = SimulationDataset(self.dataset, scene_indices, self.sim_cfg.start_frame_index,
+                                        self.sim_cfg.disable_new_agents, self.sim_cfg.distance_th_far,
+                                        self.sim_cfg.distance_th_close)
+
+        if self.sim_cfg.num_simulation_steps is None:
+            range_unroll = range(self.sim_cfg.start_frame_index, len(sim_dataset))
+        else:
+            end_frame_index = min(len(sim_dataset), self.sim_cfg.num_simulation_steps + self.sim_cfg.start_frame_index)
+            range_unroll = range(self.sim_cfg.start_frame_index, end_frame_index)
+
+        for frame_index in tqdm(range_unroll):
+            next_frame_index = frame_index + 1
+            should_update = next_frame_index != range_unroll.stop
+
+            # AGENTS
+            if not self.sim_cfg.use_agents_gt:
+                agents_input = sim_dataset.rasterise_agents_frame_batch(frame_index)
+                agents_input_dict = default_collate(list(agents_input.values()))
+                agents_output_dict = self.model_agents(agents_input_dict)
+                if should_update:
+                    self.update_agents(sim_dataset, next_frame_index, agents_input_dict, agents_output_dict)
+
+            # EGO
+            if not self.sim_cfg.use_ego_gt:
+                ego_input = sim_dataset.rasterise_frame_batch(frame_index)
+                ego_input_dict = default_collate(ego_input)
+                ego_output_dict = self.model_ego(ego_input_dict)
+                if should_update:
+                    self.update_ego(sim_dataset, next_frame_index, ego_input_dict, ego_output_dict)
+
+        simulated_outputs: List[SimulationOutputs] = []
+        for scene_idx in scene_indices:
+            simulated_outputs.append(SimulationOutputs(scene_idx, sim_dataset))
```

Comment:
I think it would be better here to pass to `SimulationOutputs` only what is needed from the `SimulationDataset` to build it, instead of the entire object and then getting what is needed from inside of the `SimulationOutputs`, otherwise, it makes it difficult to understand what it uses from the dataset just by looking at the constructor.

Reply:
yes, but right now we still don't know what we need, that's the issue :)